build_tools/python/e2e_test_framework/definitions/iree_definitions.py

```diff
@@ -95,7 +95,20 @@ class ModuleExecutionConfig(object):
   tags: List[str]
   loader: RuntimeLoader
   driver: RuntimeDriver
-  extra_flags: List[str] = dataclasses.field(default_factory=list)
+  # Acutal additional flags to be appended during the run.
+  additional_flags: List[str] = dataclasses.field(default_factory=list,
+                                                  init=False)
+  # Pseudo init field that will be appended to `additional_flags` field.
+  extra_flags: dataclasses.InitVar[List[str]] = []
+
+  def __post_init__(self, extra_flags: List[str]):
+    # Special flags that we want to enable for all benchmarks. We would like to
+    # keep these flags in the config so the exported config contains all
+    # information for reproducing. Add/remove flags with caution.
+    global_flags = ["--device_allocator=caching"]
+    # __setattr__ is required due to the `frozen=True` (
+    # https://docs.python.org/3/library/dataclasses.html#frozen-instances)
+    object.__setattr__(self, "additional_flags", global_flags + extra_flags)
```

Comment:
This has confusing naming "extra" and "additional" mean the same thing, so it doesn't make clear what the distinction is between these (one is for initialization and the other is used). But I think more generally this is overly complex. Not so much because of the dataclasses cleverness, but rather because it makes this global flag implicit. What I was suggesting was that we really should make the usage of caching allocator part of the benchmark. It should be possible to use something else, and if we do then that should have a different benchmark ID. So the starting place I would use is just including it in "extra_flags" for every config. Ok, so that's annoying, let's add a function so we don't have to repeat ourselves so much. I think where that gets us is something like a helper function that appends "--device_allocator=caching" to extra_flags. I'd suggest this:


```suggestion
  extra_flags: List[str] = dataclasses.field(default_factory=list)
  
  @classmethod
  def with_caching_allocator(cls, *, id, tags, loader, driver, extra_flags) -> "ModuleExecutionConfig":
    extra_flags = extra_flags or []
    extra_flags.append("--device_allocator=caching")
    return cls(id, tags, loader, driver, extra_flags)
```

We could call the function `with_default_flags` as well, although then there's the danger that we could want to add a new default flag. This is all just syntactic sugar though, so we can play with the code later. I think explicit is better than implicit here though.

Reply:
SG.

It's a default flag for benchmarking (at least from description in the original PR), so probably I put it in `benchmark_suites.iree.module_execution_configs` instead. If we ever want to enable it even broader, then it probably should be default in iree runtime.

Reply:
It's *currently* the default for benchmarking, but the original PR mentions that we will hopefully have benchmarks in the future where it would be the wrong choice:


> For now this means that `--device_allocator=caching` needs to be passed to benchmark tooling when used as there's not yet a good way to force it on just for that binary (or if we'd even want to - hopefully benchmarks end up having some dynamic behavior in which case the caching allocator would be the wrong one to use). Clever users wanting to more tightly control the behavior can use a somewhat esoteric syntax to configure pools:

Reply:
Then I think the right way is to have a "caching-allocator" tag on the current execution config, just to remind us all of these benchmarks are run with the caching allocator.

If we have other benchmarks w/o caching allocator, we can have the separate execution config for them (and potentially rename the current `ELF_LOCAL_SYNC_CONFIG` to `ELF_LOCAL_SYNC_CACHING_ALLOCATOR_CONFIG`). But I don't think we need to make that much noise for now (and we are free to change these names as long as the actually ids stay the same).

Reply:
Do we really need to include it in tags? One of the advantages to artificial IDs is that we no longer need tags to provide permanent unique IDs. I wouldn't add this until we have usage of non-caching allocator. In fact, I think we could probably drop "default-flags" too (though let's wait on that)

Reply:
Removed